pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java

```diff
@@ -246,78 +288,119 @@ public Object visit(ASTAnnotationTypeDeclaration node, Object data) {
+     * Searches a JavaTypeDefinition and it's superclasses until a field with name {@code fieldImage} that
+     * is visible from the {@code accessingClass} class. Once it's found, it's possibly generic type is
+     * resolved with the help of {@code typeToSearch} TypeDefinition.
+     *
+     * @param typeToSearch   The type def. to search the field in.
+     * @param fieldImage     The simple name of the field.
+     * @param accessingClass The class that is trying to access the field, some Class declared in the current ACU.
+     * @return JavaTypeDefinition of the resolved field or null if it could not be found.
+     */
+    private JavaTypeDefinition getFieldType(JavaTypeDefinition typeToSearch, String fieldImage, Class accessingClass) {
+        while (typeToSearch != null) {
+            try {
+                Field field = typeToSearch.getType().getDeclaredField(fieldImage);
+                if (isMemberVisibleFromClass(typeToSearch.getType(), field.getModifiers(), accessingClass)) {
+                    return getNextTypeDefinition(typeToSearch, field.getGenericType());
+                }
+            } catch (NoSuchFieldException e) { /* swallow */ }
+
+            // transform the type into it's supertype
+            typeToSearch = getNextTypeDefinition(typeToSearch, typeToSearch.getType().getGenericSuperclass());
+        }
+
+        return null;
+    }
+
+    /**
+     * Search for a field by it's image stating from a scope and taking into account if it's visible from the
+     * accessingClass Class. The method takes into account that Nested inherited fields shadow outer scope fields.
+     *
+     * @param scope          The scope to start the search from.
+     * @param image          The name of the field, local variable or method parameter.
+     * @param accessingClass The Class (which is defined in the current ACU) that is trying to access the field.
+     * @return Type def. of the field, or null if it could not be resolved.
+     */
+    private JavaTypeDefinition getTypeDefinitionOfVariableFromScope(Scope scope, String image, Class accessingClass) {
         if (accessingClass == null) {
             return null;
         }
 
         for (/* empty */; scope != null; scope = scope.getParent()) {
+            // search each enclosing scope one by one
             for (Map.Entry<VariableNameDeclaration, List<NameOccurrence>> entry
                     : scope.getDeclarations(VariableNameDeclaration.class).entrySet()) {
                 if (entry.getKey().getImage().equals(image)) {
-                    return entry.getKey().getType();
+                    ASTType typeNode = entry.getKey().getDeclaratorId().getTypeNode();
+
+                    if (typeNode.jjtGetChild(0) instanceof ASTReferenceType) {
+                        return ((TypeNode) typeNode.jjtGetChild(0)).getTypeDefinition();
+                    } else { // primitive type
+                        return JavaTypeDefinition.build(typeNode.getType());
```

Comment:
add a TODO here, for primitives, being `JavaTypeDefinition` immutable, it may make sense to have singleton instances to use (probably within a Map).